x/storage/keeper/keeper.go

```diff
@@ -1409,20 +1440,74 @@ func (k Keeper) UpdateGroupMember(ctx sdk.Context, operator sdk.AccAddress, grou
 
 	}
 	if err := ctx.EventManager().EmitTypedEvents(&types.EventUpdateGroupMember{
-		Operator:        operator.String(),
-		Owner:           groupInfo.Owner,
-		GroupName:       groupInfo.GroupName,
-		GroupId:         groupInfo.Id,
-		MembersToAdd:    opts.MembersToAdd,
-		MembersToDelete: opts.MembersToDelete,
+		Operator:           operator.String(),
+		Owner:              groupInfo.Owner,
+		GroupName:          groupInfo.GroupName,
+		GroupId:            groupInfo.Id,
+		MembersToAdd:       opts.MembersToAdd,
+		MembersToDelete:    opts.MembersToDelete,
+		AddedMembersDetail: addedMembersDetailEvent,
 	}); err != nil {
 		return err
 	}
 	return nil
 }
 
-func (k Keeper) UpdateGroupExtra(ctx sdk.Context, operator sdk.AccAddress, groupInfo *types.GroupInfo, extra string) error {
+func (k Keeper) RenewGroupMember(ctx sdk.Context, operator sdk.AccAddress, groupInfo *types.GroupInfo, opts types.RenewGroupMemberOptions) error {
+	if groupInfo.SourceType != opts.SourceType {
+		return types.ErrSourceTypeMismatch
+	}
+
+	// check permission
+	effect := k.VerifyGroupPermission(ctx, groupInfo, operator, permtypes.ACTION_UPDATE_GROUP_MEMBER)
+	if effect != permtypes.EFFECT_ALLOW {
+		return types.ErrAccessDenied.Wrapf(
+			"The operator(%s) has no UpdateGroupMember permission of the group(%s), operator(%s)",
+			operator.String(), groupInfo.GroupName, groupInfo.Owner)
+	}
 
+	eventMembersDetail := make([]*types.EventGroupMemberDetail, 0, len(opts.Members))
+	for i := range opts.Members {
+		member := opts.Members[i]
+		memberExpiration := opts.MembersExpiration[i].UTC()
+		memberAcc, err := sdk.AccAddressFromHexUnsafe(member)
+		if err != nil {
+			return err
+		}
+
+		groupMember, found := k.permKeeper.GetGroupMember(ctx, groupInfo.Id, memberAcc)
+		if !found {
+			err = k.permKeeper.AddGroupMember(ctx, groupInfo.Id, memberAcc, &memberExpiration)
+			if err != nil {
+				return err
+			}
+		} else {
+			k.permKeeper.UpdateGroupMember(ctx, groupInfo.Id, memberAcc, groupMember.Id, &memberExpiration)
+		}
+
+		eventMembersDetail = append(eventMembersDetail, &types.EventGroupMemberDetail{
+			Member:         member,
+			ExpirationTime: memberExpiration,
+		})
+	}
+
+	if err := ctx.EventManager().EmitTypedEvents(&types.EventRenewGroupMember{
+		Operator:      operator.String(),
+		Owner:         groupInfo.Owner,
+		GroupName:     groupInfo.GroupName,
+		GroupId:       groupInfo.Id,
+		SourceType:    groupInfo.SourceType,
+		Members:       opts.Members,
+		Extra:         groupInfo.Extra,
```

Comment:
I think this `Extra` is not needed, as it is a field of a already exist group.

Reply:
removed